File: сross_platform_software_systems/14/user_service/user_service/src/main/java/com/c3equalz/user_service/application/common/ports/EventBus.java
package com.c3equalz.user_service.application.common.ports;

import com.c3equalz.user_service.domain.common.BaseDomainEvent;
import reactor.core.publisher.Mono;

import java.util.List;

/**
 * Interface for publishing domain events.
 * <p>
 * This interface allows publishing domain events to an event bus or message broker.
 */
public interface EventBus {

    /**
     * Publishes a list of domain events.
     *
     * @param events the list of domain events to publish
     * @return Mono that completes when all events are published
     */
    Mono<Void> publish(List<BaseDomainEvent> events);
}

File: сross_platform_software_systems/14/chat_service/src/chat_service/application/common/ports/event_bus.py
from abc import abstractmethod
from collections.abc import Iterable
from typing import Protocol

from chat_service.domain.common.events import BaseDomainEvent


class EventBus(Protocol):
    @abstractmethod
    async def publish(self, events: Iterable[BaseDomainEvent]) -> None:
        raise NotImplementedError
